### ISSUE TITLE
plan failing constraint after upgrade

### ISSUE THREAD

**Issue by EugenKon** [hashicorp/nomad#22444]
### Proposal
1. It is not clear what does this mean:
![image](https://github.com/hashicorp/nomad/assets/139780495/a47b6043-33ee-4008-ab19-d2e6ed78ac17)

Eg. which drivers were missed?

I just upgraded my nomad and other tools:
```
-CONSULVERSION=1.17.3
+CONSULVERSION=1.18.2

-VAULTVERSION=1.14.9
+VAULTVERSION=1.16.2

-NOMADVERSION=1.7.5
+NOMADVERSION=1.8.0

-CONSULTEMPLATEVERSION=0.37.0
+CONSULTEMPLATEVERSION=0.37.6
```

Before upgrade it was fine. After the upgrade configuration was not changed.

2. Another issue I encountered after upgrade is: `Error during plan: Unexpected response code: 500 (No path to region)`
This was resolved just by removing `region = "xxx"`.

Why this happens?

clinet_nomad.hcl:
```
region = "xxx"
datacenter = "dc1"
...
```

service.hcl
```
job "portal" {
  region = "xxx"  # If I comment this line, then `nomad job plan service.hcl` works
  datacenters = ["dc1"]
  type = "service"
  ...
}
```

**Comment by EugenKon** [comment]
Partially related: https://github.com/hashicorp/nomad/issues/20277

**Comment by tgross** [comment]
@EugenKon I'm closing this issue as a dupe of #20277, because although it's not the same plan error it's got the same symptoms. Open a separate issue for your other issue, please.

**Comment by EugenKon** [comment]
@tgross Sorry, let me rephrase the original issue. On the previous version of Nomad the same configuration works fine, but after upgrade it failed with the error message above:
![image](https://github.com/hashicorp/nomad/assets/139780495/2e729410-2982-4bcf-bc2f-178aec32be31)

The main inquire not about filtering itself, but about that the same configuration works fine on the Nomad v1.7.5.

The changes I did:
![image](https://github.com/hashicorp/nomad/assets/139780495/e86ac123-3d16-4d61-9353-056bd734e102)
I used https://github.com/hashicorp/learn-nomad-cluster-setup to build the base AMI for my cluster.

**Comment by tgross** [comment]
@EugenKon can you post the jobspec and plan result from the CLI (not the UI)?

**Comment by EugenKon** [comment]
```
nomad job plan -var="project_name=$PRTL_PROJECT_NAME" derived-src/services/cluster.hcl
Job: "portal"
Task Group: "postgres-db" (1 create)
  Task: "postgres-task"

Task Group: "redis-db" (1 create)
  Task: "redis"

Task Group: "services" (1 create)
  Task: "api-task"
  Task: "draft-task"
  Task: "knecht-task"
  Task: "nginx-task"
  Task: "webshot-task"

Scheduler dry-run:
- WARNING: Failed to place all allocations.
  Task Group "postgres-db" (failed to place 1 allocation):
    * Constraint "missing drivers": 1 nodes excluded by filter

  Task Group "redis-db" (failed to place 1 allocation):
    * Constraint "missing drivers": 1 nodes excluded by filter

  Task Group "services" (failed to place 1 allocation):
    * Constraint "missing drivers": 1 nodes excluded by filter

Job Warnings:
2 warnings:

* Memory oversubscription is not enabled; Task "services.knecht-task" memory_max value will be ignored. Update the Scheduler Configuration to allow oversubscription.
* Memory oversubscription is not enabled; Task "services.webshot-task" memory_max value will be ignored. Update the Scheduler Configuration to allow oversubscription.

Job Modify Index: 22
To submit the job with version verification run:

nomad job run -check-index 22 -var="project_name=nomad" derived-src/services/cluster.hcl

When running the job with the check-index flag, the job will only be run if the
job modify index given matches the server-side version. If the index has
changed, another user has modified the job and the plan's results are
potentially invalid.
```

@tgross May I PM to you the `cluster.hcl` file and other configuration files?

**Comment by tgross** [comment]
> May I PM to you the cluster.hcl file and other configuration files?

You can email it to nomad-oss-debug@hashicorp.com, which is a send-only email address which only members of the Nomad engineering team can read. Include this issue number in the subject.

**Comment by tgross** [comment]
@EugenKon can you also verify that the client nodes you're targeting do have the drivers you're using and that they're healthy? You can see those with `nomad node status -verbose $node_id`

**Comment by EugenKon** [comment]
@tgross  I have sent the configuration files to the specified address.

Here is the output from `nomad node status`:
```
nomad node status -verbose fe2981b0
ID              = fe2981b0-caa2-d8f5-55ce-378dac9a6882
Name            = ip-172-31-0-133
Node Pool       = default
Class           = <none>
DC              = dc1
Drain           = false
Eligibility     = eligible
Status          = ready
CSI Controllers = <none>
CSI Drivers     = <none>
Uptime          = 64h2m31s

Drivers
Driver    Detected  Healthy  Message                             Time
docker    false     false    Failed to connect to docker daemon  2024-06-01T00:30:52Z
exec      true      true     Healthy                             2024-06-01T00:30:52Z
java      false     false    <none>                              2024-06-01T00:30:52Z
qemu      false     false    <none>                              2024-06-01T00:30:52Z
raw_exec  false     false    disabled                            2024-06-01T00:30:52Z

Node Events
Time                  Subsystem  Message          Details
2024-06-01T00:30:54Z  Cluster    Node registered  <none>

Allocated Resources
CPU         Memory       Disk
0/2400 MHz  0 B/949 MiB  0 B/4.2 GiB

Allocation Resource Utilization
CPU         Memory
0/2400 MHz  0 B/949 MiB

Host Resource Utilization
CPU          Memory           Disk
24/2400 MHz  225 MiB/949 MiB  (/dev/root)

Allocations
No allocations placed

Attributes
consul.connect                           = true
consul.datacenter                        = dc1
consul.dns.addr                          = 172.31.0.133
consul.dns.port                          = 8600
consul.ft.namespaces                     = false
consul.grpc                              = 8502
consul.revision                          = 9fc827ca
consul.server                            = true
consul.sku                               = oss
consul.version                           = 1.18.2
cpu.arch                                 = amd64
cpu.frequency.efficiency                 = 2400
cpu.frequency.performance                = 0
cpu.modelname                            = Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz
cpu.numcores                             = 1
cpu.numcores.efficiency                  = 1
cpu.numcores.performance                 = 0
cpu.reservablecores                      = 1
cpu.totalcompute                         = 2400
cpu.usablecompute                        = 2400
driver.exec                              = 1
kernel.arch                              = x86_64
kernel.landlock                          = v3
kernel.name                              = linux
kernel.version                           = 6.5.0-1020-aws
memory.totalbytes                        = 995373056
nomad.advertise.address                  = 172.31.0.133:4646
nomad.bridge.hairpin_mode                = false
nomad.revision                           = 28b82e4b2259fae5a62e2ed47395334bea5a24c4
nomad.service_discovery                  = true
nomad.version                            = 1.8.0
numa.node.count                          = 1
numa.node0.cores                         = 0
os.cgroups.version                       = 2
os.name                                  = ubuntu
os.signals                               = SIGTTIN,SIGALRM,SIGCONT,SIGHUP,SIGTSTP,SIGPROF,SIGTRAP,SIGUSR1,SIGSYS,SIGBUS,SIGINT,SIGKILL,SIGPIPE,SIGQUIT,SIGTERM,SIGFPE,SIGILL,SIGIOT,SIGUSR2,SIGNULL,SIGSTOP,SIGWINCH,SIGXCPU,SIGXFSZ,SIGABRT,SIGIO,SIGSEGV,SIGTTOU
os.version                               = 22.04
platform.aws.ami-id                      = ami-069e***
platform.aws.instance-life-cycle         = on-demand
platform.aws.instance-type               = t2.micro
platform.aws.placement.availability-zone = us-west-2a
unique.consul.name                       = ip-172-31-0-133
unique.hostname                          = ip-172-31-0-133
unique.network.ip-address                = 172.31.0.133
unique.platform.aws.hostname             = ip-172-31-0-133.us-west-2.compute.internal
unique.platform.aws.instance-id          = i-0235***
unique.platform.aws.local-hostname       = ip-172-31-0-133.us-west-2.compute.internal
unique.platform.aws.local-ipv4           = 172.31.0.133
unique.platform.aws.mac                  = 06:bc:50:ac:4d:41
unique.platform.aws.public-hostname      = ec2-****.us-west-2.compute.amazonaws.com
unique.platform.aws.public-ipv4          = ***
unique.storage.bytesfree                 = 4544131072
unique.storage.bytestotal                = 8132173824
unique.storage.volume                    = /dev/root

Meta
connect.gateway_image                           = docker.io/envoyproxy/envoy:v${NOMAD_envoy_version}
connect.log_level                               = info
connect.proxy_concurrency                       = 1
connect.sidecar_image                           = docker.io/envoyproxy/envoy:v${NOMAD_envoy_version}
connect.transparent_proxy.default_outbound_port = 15001
connect.transparent_proxy.default_uid           = 101
```

Ah, Ok. I can see from this that Nomad can not connect to Docker.

But this does not mean that driver is missing, the driver is there, Nomad can not connect to it. Should the error message be fixed?

**Comment by EugenKon** [comment]
So the problem is on the `server` instance:
![image](https://github.com/hashicorp/nomad/assets/139780495/5672b3e9-3eb6-4ea4-b30c-99866c1c4a00)

The docker is not supposed to be started there.

I am not sure why `nomad node status` returns only one node (see below) when Consul shows two nodes (see the screenshot above):
![image](https://github.com/hashicorp/nomad/assets/139780495/e12bad8c-fc07-4452-bbdf-8167c5bfca6e)

**Comment by EugenKon** [comment]
Notice that region is 'global' and not as configured: 'xxx'.

<details>
  <summary>The Nomad's log on `1.61` host</summary>

```
Jun 01 00:29:47 ip-172-31-1-61 systemd[1]: Starting Nomad...
░░ Subject: A start job for unit nomad.service has begun execution
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░
░░ A start job for unit nomad.service has begun execution.
░░
░░ The job identifier is 4703.
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> WARNING: mTLS is not configured - Nomad is not secure without mTLS!
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> WARNING: Bootstrap mode enabled! Potentially unsafe operation.
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> Loaded configuration from /etc/nomad.d/nomad.hcl
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> Starting Nomad agent...
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> Nomad agent configuration:
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:        Advertise Addrs: HTTP: 172.31.1.61:4646; RPC: 172.31.1.61:4647; Serf: 172.31.1.61:4648
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:             Bind Addrs: HTTP: [0.0.0.0:4646]; RPC: 0.0.0.0:4647; Serf: 0.0.0.0:4648
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:                 Client: true
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:              Log Level: INFO
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:                Node Id: 3a1cc710-1496-3264-f612-b96754ef36ba
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:                 Region: global (DC: dc1)
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:                 Server: true
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:                Version: 1.8.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]: ==> Nomad agent started! Log data will stream in below:
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.196Z [INFO]  nomad: setting up raft bolt store: no_freelist_sync=false
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.202Z [INFO]  nomad.raft: initial configuration: index=1 servers="[{Suffrage:Voter ID:bb620024-8b48-b61c-f2af-4c3e29931919 Address:172.31.1.61:4647}]"
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.202Z [INFO]  nomad.raft: entering follower state: follower="Node at 172.31.1.61:4647 [Follower]" leader-address= leader-id=
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.204Z [INFO]  nomad: serf: EventMemberJoin: ip-172-31-1-61.global 172.31.1.61
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.205Z [INFO]  nomad: starting scheduling worker(s): num_workers=4 schedulers=["system", "sysbatch", "service", "batch", "_core"]
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.205Z [INFO]  nomad: started scheduling worker(s): num_workers=4 schedulers=["system", "sysbatch", "service", "batch", "_core"]
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.206Z [INFO]  nomad: adding server: server="ip-172-31-1-61.global (Addr: 172.31.1.61:4647) (DC: dc1)"
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.208Z [WARN]  agent.plugin_loader: skipping external plugins since plugin_dir doesn't exist: plugin_dir=/opt/nomad/data/plugins
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.212Z [INFO]  agent: detected plugin: name=docker type=driver plugin_version=0.1.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.212Z [INFO]  agent: detected plugin: name=raw_exec type=driver plugin_version=0.1.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.212Z [INFO]  agent: detected plugin: name=exec type=driver plugin_version=0.1.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.212Z [INFO]  agent: detected plugin: name=qemu type=driver plugin_version=0.1.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.212Z [INFO]  agent: detected plugin: name=java type=driver plugin_version=0.1.0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.216Z [INFO]  client: using state directory: state_dir=/opt/nomad/data/client
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.224Z [INFO]  client: using alloc directory: alloc_dir=/opt/nomad/data/alloc
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.224Z [INFO]  client: using dynamic ports: min=20000 max=32000 reserved=""
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.235Z [INFO]  client.fingerprint_mgr.consul: consul agent is available: cluster=default
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.344Z [WARN]  client.fingerprint_mgr.network: unable to parse speed: path=/usr/sbin/ethtool device=lo
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.352Z [WARN]  client.fingerprint_mgr.network: unable to parse speed: path=/usr/sbin/ethtool device=docker0
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.354Z [WARN]  client.fingerprint_mgr.cni_plugins: failed to read CNI plugins directory: cni_path=/opt/cni/bin error="open /opt/cni/bin: no such file or directory"
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.375Z [INFO]  client.proclib.cg2: initializing nomad cgroups: cores=0-3
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.375Z [INFO]  client.plugin: starting plugin manager: plugin-type=csi
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.375Z [INFO]  client.plugin: starting plugin manager: plugin-type=driver
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.376Z [INFO]  client.plugin: starting plugin manager: plugin-type=device
Jun 01 00:29:55 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:55.397Z [INFO]  client: started client: node_id=7744b45b-464a-3f66-2e89-2e1d3c19f1c6
Jun 01 00:29:55 ip-172-31-1-61 systemd[1]: Started Nomad.
░░ Subject: A start job for unit nomad.service has finished successfully
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░
░░ A start job for unit nomad.service has finished successfully.
░░
░░ The job identifier is 4703.
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.180Z [WARN]  nomad.raft: heartbeat timeout reached, starting election: last-leader-addr= last-leader-id=
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.180Z [INFO]  nomad.raft: entering candidate state: node="Node at 172.31.1.61:4647 [Candidate]" term=2
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.188Z [INFO]  nomad.raft: election won: term=2 tally=1
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.188Z [INFO]  nomad.raft: entering leader state: leader="Node at 172.31.1.61:4647 [Leader]"
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.188Z [INFO]  nomad: cluster leadership acquired
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.308Z [INFO]  nomad.core: established cluster id: cluster_id=88fd5f54-8443-9c19-0311-a451b46a5cc7 create_time=1717201797299270033
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.308Z [INFO]  nomad: eval broker status modified: paused=false
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.308Z [INFO]  nomad: blocked evals status modified: paused=false
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.383Z [INFO]  client: node registration complete
Jun 01 00:29:57 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:29:57.616Z [INFO]  nomad.keyring: initialized keyring: id=48a17586-46ff-2bfd-3260-7ad22f0c70c3
Jun 01 00:30:05 ip-172-31-1-61 nomad[1933]:     2024-06-01T00:30:05.134Z [INFO]  client: node registration complete
^C
ubuntu@ip-172-31-1-61:~$ exit
logout
Connection to 35.90.43.81 closed.
kes@Eugens-MacBook-Pro nomad $ nomad status
ID      Type     Priority  Status   Submit Date
portal  service  50        pending  2024-05-31T20:41:44-04:00
kes@Eugens-MacBook-Pro nomad $ nomad node status
ID        Node Pool  DC   Name             Class   Drain  Eligibility  Status
fe2981b0  default    dc1  ip-172-31-0-133  <none>  false  eligible     ready
```
</details>

**Comment by EugenKon** [comment]
@tgross It seems I can not send to that email:
![image](https://github.com/hashicorp/nomad/assets/139780495/39ca1838-f63e-4342-aef8-ca92aabe7c32)

**Comment by EugenKon** [comment]
So there is [the same error behaviour](https://github.com/hashicorp/nomad/issues/22444#issuecomment-2145550599) in 1.7.7

The latest version where my configuration works is 1.7.6:

![image](https://github.com/hashicorp/nomad/assets/139780495/b611e72a-6499-4ad1-8d9c-0489f0f08c5a)


```
$ ubuntu@ip-172-31-14-8:~$ docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
ubuntu@ip-172-31-14-106:~$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
```

![image](https://github.com/hashicorp/nomad/assets/139780495/f05a7b6c-1f0c-41d4-9154-87e9ef20dc3f)

![image](https://github.com/hashicorp/nomad/assets/139780495/eacf1adf-3410-4369-9853-649c12670c0b)

**Comment by tgross** [comment]
@EugenKon we're a bit all over the place here... let me summarize my understanding of the scenario at this point.

* Agent 172.31.14.106:
  * This agent is a combined Nomad server + Nomad client (reminder: this is not recommended)
  * This agent is not running Docker
  * This agent has registered itself as a Nomad client service with Consul 
  * This agent has registered itself as a Nomad client `fe2981b0` with the Nomad server (itself)
* Agent 172.31.14.8
  * This agent is only a Nomad client
  * This agent is running Docker
  * This agent has registered itself as a Nomad client service with Consul
  * This agent has _not_ registered itself as a Nomad client with the Nomad server
  * If you downgrade the agent to Nomad 1.7.6, it does successfully register itself with Nomad

Therefore:
* Scheduling a job with `docker` tasks fails, because the only client with Docker isn't registered with Nomad

You've shared info-level logs from agent 172.31.1.61, which is a combined Nomad server + Nomad client, but it's not clear how that's relevant.

The way to debug what's happening here is to look at debug-level logs for the agent 172.31.14.8, which is running as a client where you expect to be able to run workloads. It's not registering itself with the Nomad servers, but I can't tell why without those logs. Please share those logs (text, not screenshots please).

---

> @tgross It seems I can not send to that email

Sorry, not sure what the issue is there. That's a Google Group and the security settings say "Anyone on the web: can post". It probably doesn't matter given that we've narrowed down this problem to the node not registering itself. The logs I've asked for will be sufficient to determine what to debug next.


> But this does not mean that driver is missing, the driver is there, Nomad can not connect to it. Should the error message be fixed?

The error message is very intentional. The scheduler is saying that when it checked feasibility of that node, the driver was not present in the node's fingerprint. The `message` field of the fingerprint isn't in the scope of things the scheduler is supposed to be looking at; it just gets a "yes/no".

**Comment by EugenKon** [comment]
I found the guilty:
The problem was that deployment script failed, but the provisioning did not return any errors. (Are there a way to fail instance creation in case `user-data` script has non-zero exit status?)
As result `/etc/nomad.d/nomad.hcl` file was not replaced and its content was left intact:
```
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: BUSL-1.1

# Full configuration options can be found at https://developer.hashicorp.com/nomad/docs/configuration

data_dir  = "/opt/nomad/data"
bind_addr = "0.0.0.0"

server {
  # license_path is required for Nomad Enterprise as of Nomad v1.1.1+
  #license_path = "/etc/nomad.d/license.hclic"
  enabled          = true
  bootstrap_expect = 1
}

client {
  enabled = true
  servers = ["127.0.0.1"]
}
```
This is why we have seen `nomad-client` is registered from the instance designed for 'server' purposes.

 Aside question: if this is not recommended, why `nomad.hcl` is configured by default to act as server and client?

**Comment by tgross** [comment]
> (Are there a way to fail instance creation in case `user-data` script has non-zero exit status?)

Not that I'm aware of, but that's out of scope for Nomad and something you'd want to bring up with AWS support.

> Aside question: if this is not recommended, why `nomad.hcl` is configured by default to act as server and client?

By default [`server.enabled=false`](https://developer.hashicorp.com/nomad/docs/configuration/server#enabled) and [`client.enabled=false`](https://developer.hashicorp.com/nomad/docs/configuration/client#enabled). If the package had only the default behavior, the agent would not start. Generally speaking we try to include configuration files that allow the agent to start so that first time users can get started and play around with it. Whereas users deploying to production will have their own configuration management process.

It looks like this issue is resolved, so I'm going to close this issue out. I'm going to ask that for further questions you have with that tutorial repo that you please open issues in that repo specifically so that the folks who own that work can reply. Thanks.
